### ISSUE TITLE
Using disparity images/maps to get real distance information

### ISSUE THREAD

**Issue by yusiyoh** [julimueller/dtld_parsing#14]
Hello, 
Firstly, thank you for your amazing work. I am currently working on DTLD for my thesis. As mentioned in the title, I need to calculate real distance by using disparity images. Finding the depth is straightforward with focal length, baseline, and disparity matrix (considering parallel camera setup). However, I could not figure out how to calculate the 3D location of a point from the depth information. In OpenCV, there is the function cv2.reprojectImageTo3D(disparity_image, Q) where it requires 4x4 Q matrix (projection transformation matrix) and according to documentation, it can be obtained by cv2.stereoRectify(), but this one requires the following: 

cameraMatrix1 – First camera matrix.
cameraMatrix2 – Second camera matrix.
distCoeffs1 – First camera distortion parameters.
distCoeffs2 – Second camera distortion parameters.
imageSize – Size of the image used for stereo calibration.
R – Rotation matrix between the coordinate systems of the first and the second cameras.
T – Translation vector between coordinate systems of the cameras.

Checking the calibration data published, I see that I need intrinsic matrix and distortion coefficients for the right camera too.
Could you help me about this task? Thank you very much.

**Comment by yusiyoh** [comment]
While checking the calibration data, I noticed that the projection matrix has a null 4th column. This means it has whether 0 focal length or baseline which is not possible I guess.

**Comment by yusiyoh** [comment]
**UPDATE:** Instead of using cv2.stereoRectify() to get the Q matrix, I manually constructed it according to the documentation of cv2. However, before calculating the distance, I checked the depth values for a sample image /Bochum/Bochum1/2015-04-21_17-11-13/DE_BBBR667_2015-04-21_17-11-15-686220_k0.tiff: 
![image](https://user-images.githubusercontent.com/6302128/145671379-76fa5e95-64c4-43c0-9a0f-5d9fb615e00f.png)

The calculate the depth, I used the basic formula: focal * baseline / disparity
I got the disparity image by using your parser function: https://github.com/julimueller/dtld_parsing/blob/065f540a9e5ddfca8b69e92800d645c9a1364d85/python/dtld_parsing/driveu_dataset.py#L194-L220
After getting the image, I doubled the disparity values since in the documentation it is stated so:
![image](https://user-images.githubusercontent.com/6302128/145671526-459900a0-a7b6-40df-9651-06bb1de55a85.png)
The resulting depth values are as follows which seems not correct since there are many 500-600m readings:
![image](https://user-images.githubusercontent.com/6302128/145671549-93e4e697-7207-4c54-86ef-1f50b9561fb0.png)
Then, I eliminated the 1/16 scaling in the parser function and doubled the values as before to see the change and got this result: 
![image](https://user-images.githubusercontent.com/6302128/145671601-6d073f8c-c1dd-422d-adb2-de37d6d9099d.png)

This time, the depth values seem to be wrong again since they are around 40m max (and of course some inf readings). 
Can you clarify how to use the disparity image in the dataset to correctly calculate the depth and more importantly distance for left images? 
Thank you very much.

**Comment by julimueller** [comment]
Hi @yusiyoh 

I know that OpenCV 3D Processing can be a pain. I recommend to take a look into the ROS pinhole camera model and stereo camera model libraries. 

http://docs.ros.org/en/indigo/api/image_geometry/html/c++/stereo__camera__model_8cpp_source.html

The method projectDisparityTo3d shows how to project a rectified 2D point to 3D. The required data you need should bei available, left and right intrinsics, baseline. In order to get the disparity value just use map_labels_to_disparity_image method in dtld parsing to get the object box in the disparity image (be aware of the downsampled disparity image!) and then average/median the disparity values of the box. The rectified points needed for projectDisparityTo3d can be obtained similar to map_labels_to_disparity_image but without downsampling by the factor of 2 (Basically just use undistortpoints).

Let me know if this helps you!

**Comment by yusiyoh** [comment]
Thank you very much for your answer. I will proceed as you proposed. I can share the outcome here if you want to see and approve.

**Comment by julimueller** [comment]
Sure, let me know! If it does not work I will try to extend dtld_parsing by a method returning the 3D pos for each labeled box.

**Comment by yusiyoh** [comment]
I have created a colab notebook and tried to create a function like projectDisparityTo3d. I used the parser function again as I did before. The outcome is different from before for x,y coordinates but z values (depth) are the same, which are still very high to be correct I guess. 
You can check the notebook here: https://colab.research.google.com/drive/1KmooWiAMGW4k6suW565PllFacU1j43aV?usp=sharing
I used the functions in the dtld_parsing and after getting mean disparities for each traffic light, I scaled the undistorted points by 2 in the projectDisparityTo3d I defined. 
I think that I am missing some points. Thank you for your help, waiting for your response.

**Comment by yusiyoh** [comment]
Dear Sir @julimueller, 

Did you have a chance to look at the notebook? I am a bit stuck and cannot move on about this part of my study. If you can have a look at it, it might be solved faster, and then I may try to implement it as PR for this repo. 
Thanks again. Waiting for your response.

**Comment by julimueller** [comment]
hi @yusiyoh,

I added a functionality to obtain the 3D position for each label. Take a look and give me some feedback if this does what you want!

Merry christmas,

Julian

**Comment by yusiyoh** [comment]
Dear Sir @julimueller,

I checked the code and run the script. This is exactly what I needed. In addition, I found my mistake while checking your new script. I multiplied the undistorted points with 2 (binning factor) but forgot to multiply disparity values which was the whole problem. Actually, it could have been solved 11 days ago :(  Of course, your code is more elegant and properly implemented. 

Thank you very much for your help and fast responses. The next step for me will be tracking them by implementing KF, which I hope to solve asap. I can share the results with you if it is any good for this repo. It was a pleasure to talk with you.

Merry Christmas,
Yusuf Can
